### ISSUE TITLE
Asciidoc task leaves file handle to processed asciidoc file open

### ISSUE THREAD

**Issue by jjohannes** [asciidoctor/asciidoctor-gradle-plugin#214]
I encountered this issue in the [gradle/gradle](https://github.com/gradle/gradle) build. When running the build in a Gradle daemon on Windows, file handles to the input files are still open after the build finished. I assume that the files opened by the asciidoc tool are not properly closed.

The workaround for us is to replace the `AsciidoctorProxyImpl.renderFile()` code with code that first pulls the content from the file and then only passes the content in using `delegate.render()` instead of `delegate.renderFile()`.

https://github.com/gradle/gradle/blob/master/subprojects/docs/docs.gradle#L130-L134

This solution works for our build but is not generalisable, because the output file path depends on the backend used (we only use XML). I am not sure what the best general fix would be as I am not familiar with the underlying code base.

**Comment by mojavelinux** [comment]
Until someone can prove this this is Asciidoctor leaving the files open, I'm going to assume it's a bug in Gradle itself. If it is Asciidoctor, we will move immediately on a fix. But I just don't see where this is happening.

**Comment by jjohannes** [comment]
Unfortunately this is hard to reproduce in a test case. And I am a total Ruby noob.

Might this `open()` here need a `close()`?
https://github.com/asciidoctor/asciidoctor/blob/08ffb19a544d26b14a6c210706399e4f0d798037/lib/asciidoctor.rb#L1377

**Comment by mojavelinux** [comment]
Nope. The block form is self closing. See http://ruby-doc.org/core-2.4.1/IO.html#method-c-open

**Comment by mojavelinux** [comment]
> The workaround for us is to replace the AsciidoctorProxyImpl.renderFile() code with code that first pulls the content from the file and then only passes the content in using delegate.render() instead of delegate.renderFile().

The other problem with this approach is that many things don't get set when you do it this way.

(FYI, render is deprecated...it's now called convert).

**Comment by mojavelinux** [comment]
Are you using Asciidoctor 1.5.6?

**Comment by mojavelinux** [comment]
My mistake. I was wrong. Asciidoctor core (or AsciidoctorJ) may indeed be leaving the file handle open in the case used by the Gradle plugin. This situation occurs when a File object is passed to convert_file instead of a String path. Core assumes this object is being managed by the caller and therefore only reads from it.

We first need to decide who's responsibility it is to close it. It's not necessarily the plugin's job. It might be core after all. Then we can make the necessary change.

Of course, the plugin could close the file after calling convertFile in the meantime.

**Comment by mojavelinux** [comment]
I was wrong again. AsciidoctorJ never passes a File object to Asciidoctor Ruby. (In fact, if it did, it would fail). Instead, it calls getAbsolutePath() on the File before invoking the underlying convert_file method.

So I'm back to not knowing how it could possibly be leaving a file open.

Is there anyway you can trace when files are opened and closed? We really need more information.

**Comment by k-mack** [comment]
@jjohannes I tried to trace this from outside of Gradle using [File Leak Detector](http://file-leak-detector.kohsuke.org/). I posted the relevant parts of the captured log on the AsciidoctorJ's repository: https://github.com/asciidoctor/asciidoctorj/issues/570#issuecomment-329639859. After running the `asciidoctor` task, the source Asciidoctor file is opened, closed, and opened again. Not sure if this indicates a problem with the Gradle daemon or the asciidoctor-gradle-plugin. Can you take a look and see if anything in the stacktrace points to one or the other?

**Comment by jjohannes** [comment]
@k-mack thanks for looking into this and sorry for not getting back to this earlier.

By chance I had a Windows machine around today and ran some more tests. Actually, the issue seems to be gone with _asciidoctorj 1.5.6_. My bad @mojavelinux for not trying this before starting this whole topic. I have only checked the version of _asciidoctor-gradle-plugin_ which has not been upgraded to `1.5.6` back then.

So the underlying problem has either been fixed in aciidoctorj or one of the dependencies that upgraded with it:

```
org.asciidoctor:asciidoctorj:1.5.6 <- 1.5.3.2
org.asciidoctor:asciidoctorj-groovy-dsl:1.0.0.Alpha2 <- 1.0.0.preview2
org.jruby:jruby-complete:1.7.26 <- 1.7.21
```

Maybe it is the fix you mentioned here: https://github.com/asciidoctor/asciidoctorj/issues/570#issuecomment-321414361

I will upgrade the _gradle_ build and remove the workaround. If I run into any issues again I will let you know. But I think you can close this issue.

@k-mack if  `1.5.6` does not solve this issue for you, maybe it is something else in your case. Are you sure you tried `1.5.6` only on a fresh daemon? (If the handles are already open because you ran with an older version before, it would not release them.)

**Comment by k-mack** [comment]
@jjohannes thanks for the update. You're right -- `1.5.6` does solve this issue for the HTML and DocBook backends. I've been dealing with this issue for a while even after telling the plugin to use `asciidoctorj:1.5.6` via the plugin's `asciidoctorj {}`. (I currently have `asciidoctorj { version = '1.6.0-alpha.6' }`.

Since my asciidoctor source files are still being locked, you've made me realize that the source of my issue is different than yours. Mine must stem from using the PDF backend with `org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.15`, not this plugin. I'll have to open a separate issue elsewhere.

As far as I can tell, upgrading to `asciidoctor-gradle-plugin:1.5.6` will resolve this issue. My issue with the PDF backend is most likely isolated to `asciidoctorj-pdf`.

**Comment by ysb33r** [comment]
Looks like this has been solved in 1.5.6. Closing.
